Add version information to default user-agent header

gotchazipc opened the pull request:
PR resolves #267 

Motivation:
 - Hard to distinguish client version in server, generally need for statistics.
 - Sometimes, the version information is quite useful for client/server apps.
  (e.g. negotiating capabilities, work-around with known bugs, ...)

Modifications:
 - Added a groovy script to generate a Java class with project name & version
 - Fixed some tests that rely on hard-coded UA string

Result:
 - Now UA string is `${project.name}/${project.version}`
 - The version information is accessible as public constant string.

anuraaga commented on src/build/create-buildinfo.groovy:
```diff
@@ -0,0 +1,56 @@
+/*
```
What do you think about using the git commit I'd plugin instead of this groovy and we can also include the git hash in the user agent (can be useful for snapshot users)?

Reading the properties will be slightly more complicated than just referencing this class though.

https://github.com/ktoso/maven-git-commit-id-plugin

gotchazipc replied:
How about https://github.com/alx3apps/jgit-buildnumber plugin? maven-git-commit-id plugin looks too complex and hard to configure.

trustin commented:
How about copying some from this to include version properties file:

https://github.com/netty/netty/blob/af632278d2b966fcb45327a79345b567b56ab02c/pom.xml#L979

and ..

https://github.com/netty/netty/blob/af632278d2b966fcb45327a79345b567b56ab02c/common/src/main/java/io/netty/util/Version.java

Will be useful when we migrate to multi-module structure.

gotchazipc commented:
I'd prefer not to use resource file, because loading it  sometimes fails, due to clumsy class loaders or some other unexpected reasons. And, to move to gradle easily, I wanted to generate a source file compatible to [gradle-buildconfig](https://plugins.gradle.org/plugin/de.fuerstenau.buildconfig) plugin. 

Anyway, changing the groovy script to save information into a property file (rather than a source file)  seems to be not so hard. I'll try it, changing Netty's ant task into groovy script.

gotchazipc commented:
I have rewritten (and changed name) the groovy script, to create versions.properties. New script could be used to write the version information of each module onto single version.properties file.

trustin commented on src/build/versions-props.groovy:
```diff
@@ -0,0 +1,105 @@
+    propsFile.createNewFile();
+}
+
+// save properties to the file, only if the value of property is not falsy.
```
false

trustin commented on src/build/versions-props.groovy:
```diff
@@ -0,0 +1,105 @@
+//
+
+// Oddly, some properties are not accessible with project object.
+// Need manual configuration to access them, using global properties object.
```
Could you elaborate on this? It seems like all properties are being accessed via the 'project' object.

trustin commented on src/build/versions-props.groovy:
```diff
@@ -0,0 +1,105 @@
+
+def PROJECT_ARTIFACT_ID = project.artifactId
+def PROPS_DIR_PATH = project.build.outputDirectory + "/META-INF"
+def PROPS_FILE_PATH = PROPS_DIR_PATH + "/" + project.groupId +".versions.properties";
```
missing space after `+`

trustin commented on src/build/versions-props.groovy:
```diff
@@ -0,0 +1,105 @@
+    buildDate: PROJECT_ARTIFACT_ID + ".buildDate",
+    longCommitHash : PROJECT_ARTIFACT_ID + ".longCommitHash",
+    shortCommitHash : PROJECT_ARTIFACT_ID + ".shortCommitHash",
+    commitDate : PROJECT_ARTIFACT_ID + ".commitDate", //
```
`//` could be removed.

trustin commented on src/build/versions-props.groovy:
```diff
@@ -0,0 +1,105 @@
+// default values are taken from Netty.
+def valueOf = [
+    version : project.version,
+    buildDate : new Date().format("yyyy-MM-dd HH:mm:ss Z"), // same to git log date format
```
same with

trustin commented on src/main/java/com/linecorp/armeria/client/http/HttpHeaderUtil.java:
```diff
@@ -31,5 +34,14 @@ static String hostHeader(String host, int port, int defaultPort) {
+    private static String createUserAgentName() {
+        Version version = Version.identify().get(CLIENT_ARTIFACT_ID);
+        if (version == null) {
+            return CLIENT_ARTIFACT_ID;
```
`CLIENT_ARTIFACT_ID + "/unknown"`

trustin commented on src/main/java/com/linecorp/armeria/client/http/HttpHeaderUtil.java:
```diff
@@ -31,5 +34,14 @@ static String hostHeader(String host, int port, int defaultPort) {
+        if (version == null) {
+            return CLIENT_ARTIFACT_ID;
+        } else {
+            return CLIENT_ARTIFACT_ID + "/" + version.artifactVersion();
```
`"/"` -> `'/'`

trustin commented on src/main/java/com/linecorp/armeria/common/util/Version.java:
```diff
@@ -0,0 +1,217 @@
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
```
Could you add the Netty copyright here, since we forked from it?

trustin replied:
For example: https://github.com/line/armeria/blob/master/src/main/java/com/linecorp/armeria/internal/http/ArmeriaHttpUtil.java

trustin commented on src/main/java/com/linecorp/armeria/common/util/Version.java:
```diff
@@ -0,0 +1,217 @@
+/**
+ * Retrieves the version information of available Armeria artifacts.
+ * <p>
+ * This class retrieves the version information from {@code META-INF/com.linecorp.armeria.versions.properties},
```
```
 * <p>
 * This...
```
to:
```
 *
 * <p>This...
```

trustin commented on src/main/java/com/linecorp/armeria/common/util/Version.java:
```diff
@@ -0,0 +1,217 @@
+ * which is generated in build time. Note that it may not be possible to retrieve the information completely,
+ * depending on your environment, such as the specified {@link ClassLoader}, the current
+ * {@link SecurityManager}.
+ * </p>
```
A closing `<p>` tag causes Javadoc to fail - please remove it. http://stackoverflow.com/questions/5260368/which-tag-should-be-used-as-paragraph-separator-in-javadoc

trustin commented:
Great job! Please address the comments. I think we are pretty much done, so please squash your commits next time. :-)

trustin commented:
Merged. Thanks, @gotchazipc !

trustin commented:
Found a few issues with the PR and pushed the fixes: 9fe4c32d6d569d3edbdb04dd057caa427057f985 and 083ca166c9212c179520774b0b963dc788a02479
